control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java

```diff
@@ -35,58 +39,62 @@
     public void handlePost(ActionParameters params) throws ActionException {
         params.requireLoggedInUser();
 
-        JSONObject jsonPayload = params.getHttpParamAsJSON("featureData");
-        OskariLayer layer = getLayer(jsonPayload.optString("layerId"));
-
-        if (!canEdit(layer, params.getUser())) {
-            throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
-        }
+        try {
+            JSONArray paramFeatures = new JSONArray(params.getHttpParam("featureData"));
+            JSONArray updatedFeatureIds = new JSONArray();
+            for (int i = 0; i < paramFeatures.length(); i++) {
+                JSONObject featureJSON = paramFeatures.getJSONObject(i);
+                OskariLayer layer = getLayer(featureJSON.optString("layerId"));
+
+                if (!canEdit(layer, params.getUser())) {
+                    throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
```

Comment:
Should we check this before for each of the features/layers first? Say I try inserting five features in one request, the first four are OK but for the fifth I don't have the edit permission for that layer. The response I get is that I don't have a permission, but the first four actually completed successfully.

Reply:
Consider solving both of these issues by first iterating the `paramFeatures` array once collecting the `OskariLayer`s into a `Map<String, OskariLayer>`. Then iterate over the values of the map checking that user has edit permission for all of the layers. Then iterate over `paramFeatures` again (but this time get the `OskariLayer` from the `Map<String, OskariLayer>`) and send the WFS-T requests. Finally iterate over the values of the `Map` once again flushing the layerTilesCache once for each layer.

Reply:
Same thing for `SaveFeatureHandler`. `DeleteFeatureHandler` doesn't have the same problem as it only allows one feature per request.